### DISCUSSION TITLE
Translate tidal drum pattern modules to strudel text files

### DISCUSSION THREAD

**Discussion by urswilke** [tidalcycles/strudel#373] [Show and tell]
I've written a small script to translate the patterns in https://github.com/lvm/tidal-drum-patterns to text files for strudel. The results are [here](https://github.com/urswilke/strudel/tree/main/my-patterns).

Just wanted to let you know if it's of interest for anyone. I also tried to deploy them to github.io with the help of the instructions [here](https://github.com/tidalcycles/strudel/tree/main/my-patterns#my-patterns), but at the moment it doesn't seem to work.

**Comment by felixroos** [comment]
oh this is great, what a huge collection! It would be interesting to know which sample names are possible across all patterns, so there could exist a set of samples that can play them all. Also, they could be made available with something like this:

```js
preset('Afro1a')
//or
preset('Afro1a.bd')
```

Here is a prototype of that idea: https://strudel.tidalcycles.org?-BExVonTkS8K

I think it would be generally interesting to have a selection of preset patterns ready to use (maybe even an API to define presets, similar to `samples`)

> I also tried to deploy them to github.io with the help of the instructions [here](https://github.com/tidalcycles/strudel/tree/main/my-patterns#my-patterns), but at the moment it doesn't seem to work.

the upstream should have fixed that now

**Reply by urswilke** [reply]
> the upstream should have fixed that now

yes it worked now. See the results [here](https://urswilke.github.io/strudel/swatch/)

> It would be interesting to know which sample names are possible across all patterns, so there could exist a set of samples that can play them all. 

Sorry, I don't understand what you mean. 

Some more thoughts:

- I think your idea
  > ```js
  > preset('Afro1a')
  > //or
  > preset('Afro1a.bd')
  > ```
    is really great! I just thought my approach could help less advanced folks like me to explore strudel and patterns...

- It seems that strudel partly uses different names for the samples than that repo with the modules. Here is the list how I translated all the occurring sample names:
  |name in repo |strudel_name |
  |:----|:------------|
  |bd   |bd           |
  |cb   |cb           |
  |cow  |cb           |
  |cl   |cp           |
  |cp   |cp           |
  |ch   |hh           |
  |ht   |ht           |
  |lt   |lt           |
  |mt   |mt           |
  |oh   |oh           |
  |rm   |rim          |
  |sn   |sd           |
  |ac   |???           |
  |cy   |???           |
  |sh   |???           |
  
- I could also use a different method, like 
  ```js
  "[x ~ x ~]".s("bd")
  ```
  instead of 
  ```js
  "[bd ~ bd ~]".s()
  ```
- If you're interested, I'd be happy to make a pull request to deploy these beats into another github.io website that's directly related to strudel, instead of me (feels a bit weird to have them in a folder called "my-patterns" :sweat_smile:)

Anyways, thanks a lot for this awesome work! And please let me know if I can help (unfortunately, I know very little javascript).

**Reply by felixroos** [reply]
>>  It would be interesting to know which sample names are possible across all patterns, so there could exist a set of samples that can play them all.

> Sorry, I don't understand what you mean.

> The list of all samples occurring in these patterns are:

> `bd, ch, rm, ac, lt, mt, sn, ht, cy, oh, cp, cb, cow, cl & sh`

This answers my question :) Though I don't quite understand where those names come from, because the [Dirt-Samples](https://github.com/tidalcycles/Dirt-Samples) do not have all of these. To be specific `ch rm ac cy oh cow cl sh` are not there.. Maybe there is another set of samples I am not aware of?

> If I understand correctly, strudel partly uses different names for the samples than tidal.

This is true for the default sample set. You can open the "samples" tab in the repl to see all the default samples.
But you can also just load the Dirt-Samples [like this](https://strudel.tidalcycles.org/#c2FtcGxlcygnZ2l0aHViOnRpZGFsY3ljbGVzL0RpcnQtU2FtcGxlcy9tYXN0ZXIvJykKCnMoImJkKjIsIH4gc2QsIGhoMjcqMyxqdmJhc3MiKQ%3D%3D)

> Is there somewhere a list, how all these translate? Then I could directly use the correct names

I would not translate them, but rather load the appropriate sample list instead. 

> I could also use a different method, like ..

Another way to export this (a little more similar to how the haskell version) would be

```js
export const Afro1a = {
  bd: "[t ~ ~ ~] [~ ~ ~ ~] [t ~ ~ ~] [~ ~ ~ ~]",
  ch: "[t ~ t t] [t ~ t ~] [t ~ t ~] [t ~ t ~]",
  rm: "[~ ~ ~ ~] [~ ~ t ~] [~ ~ t ~] [~ ~ t ~]",
}
```

this would export the patterns as raw strings instead of patterns, probably making them easier to import / export. Though the chosen format depends on the way these things are intended to be used...

I'd say it would be a good start to have a matching sample list and keep the format for now.

> If you're interested, I'd be happy to make a pull request to deploy these beats into another github.io website that's directly related to strudel, instead of me

Instead of a separate repo with these, we could also just integrate them into the examples, which are currently using the same component as the my-patterns page: https://strudel.tidalcycles.org/examples/
Instead of having just one examples page, it could make sense to add multiple subpages for different purposes, where the tidal-drum-patterns would be one of them. Something like `https://strudel.tidalcycles.org/examples/tidal-drum-patterns`. It should also be useful to have an "Examples" nav link in the docs where those can be browsed.

At a later point we could then think about the preset idea, which is also related to idea b here https://github.com/tidalcycles/strudel/issues/233

**Reply by urswilke** [reply]
I edited my comment above before you were replying - sorry, if this causes confusion...

> Though I don't quite understand where those names come from, because the [Dirt-Samples](https://github.com/tidalcycles/Dirt-Samples) do not have all of these. To be specific `ch rm ac cy oh cow cl sh` are not there.. Maybe there is another set of samples I am not aware of?

I [contacted](https://github.com/lvm/tidal-drum-patterns/issues/6) the author. Probably, these were his own samples...

> I would not translate them, but rather load the appropriate sample list instead.

With my edit, I did translate most of  the unknown samples to Dirt-Samples (see table above). 
 
> we could also just integrate them [the patterns] into the examples ...

I think this would be nice! Would be happy to contribute to the documentation, if you have ideas how I can do so.

**Reply by felixroos** [reply]
Strudels default samples include https://github.com/ritchse/tidal-drum-machines#drum-names-abbreviations . I just noticed the author (@ritchse) also created an [issue about that in the tidal-drum-patterns repo](https://github.com/lvm/tidal-drum-patterns/issues/4). So it might make sense to rename them..

This is a combination of ritchses table and your mapping:

| drum                                | abbreviation | tidal-drum-patterns |
| ----------------------------------- | ------------ | ------------------- |
| Bass drum, Kick drum                | bd           | bd                  |
| Snare drum                          | sd           | **sn**              |
| Rimshot                             | rim          | **rm**              |
| Clap                                | cp           | cp                  |
| Closed hi-hat                       | hh           | **ch**              |
| Open hi-hat                         | oh           | oh                  |
| Crash                               | cr           | ? **cy** ?          |
| Ride                                | rd           | ? **cy** ?          |
| Shakers (and maracas, cabasas, etc) | sh           | **sh**              |
| High tom                            | ht           | ht                  |
| Medium tom                          | mt           | mt                  |
| Low tom                             | lt           | lt                  |
| Cowbell                             | cb           | cb                  |
| Tambourine                          | tb           | -                   |
| Other percussions                   | perc         | -                   |
| Miscellaneous samples               | misc         | -                   |
| Effects                             | fx           | -                   |
| **Accent**                          | ?            | ac                  |
| **Cymbal**                          | ?            | cy                  |


the diffs are marked in bold.

so basically, tidal-drum-patterns uses cy instead of rd/cr + has `ac` for accents, which seems agnostic to a specific sound.

It would be interesting to know where the tidal-drum-patterns come from, to know more about how it actually sounds in reality..

> I think this would be nice! Would be happy to contribute to the documentation, if you have ideas how I can do so.

If you want, you can try to add the examples subpage with the tidal-drum-patterns..
alternatively, there is [this issue](https://github.com/tidalcycles/strudel/issues/58) for undocumented functions.
docs about the docs are here : https://strudel.tidalcycles.org/technical-manual/docs/
I am also open for any suggestions / ideas / things you'd like to add to strudel

**Comment by yaxu** [comment]
Just adding a thought that what would really give these value for me, is if they could work something like editor snippets where the words would expand to mininotation sequences for editing.

**Reply by felixroos** [reply]
that would be amazing

**Reply by urswilke** [reply]
Do you mean something like [this](https://codemirror.net/examples/autocompletion/)? 

It should be straightforward to generate the code for the snippets from the tidal-drum-patterns. But regarding the javascript code in strudel, I have no idea how to implement it.
